cachi2/core/package_managers/pip.py

```diff
@@ -1370,20 +1453,131 @@ def _split_hashes_from_options(cls, options: list[str]) -> tuple[list[str], list
         return hashes, reduced_options
 
 
+def _process_req(
+    req: PipRequirement,
+    requirements_file: PipRequirementsFile,
+    require_hashes: bool,
+    pip_deps_dir: RootedPath,
+    download_info: dict[str, Any],
+    dpi: Optional[DistributionPackageInfo] = None,
+) -> dict[str, Any]:
+    download_info["kind"] = req.kind
+    download_info["requirement_file"] = str(requirements_file.file_path.subpath_from_root)
+    download_info["hash_verified"] = False
+
+    def _hash_verify(path: Path, checksum_info: Iterable[ChecksumInfo]) -> None:
+        try:
+            must_match_any_checksum(path, checksum_info)
+        except PackageRejected:
+            path.unlink()
+            log.warning("Download '%s' was removed from the output directory", path.name)
+
+    if dpi:
+        if dpi.should_verify_checksums():
+            _hash_verify(dpi.path, dpi.checksums_to_verify)
+            download_info["hash_verified"] = True
+
+        if dpi.package_type == "sdist":
+            _check_metadata_in_sdist(dpi.path)
+
+    else:
+        if require_hashes or req.kind == "url":
+            hashes = req.hashes or [req.qualifiers.get("cachito_hash", "")]
+            _hash_verify(download_info["path"], list(map(_to_checksum_info, hashes)))
+            download_info["hash_verified"] = True
+
+    log.info(
+        "-- Successfully processed '%s' in path '%s'",
+        req.download_line,
+        download_info["path"].relative_to(pip_deps_dir.root),
+    )
+
+    return download_info
+
+
+def _process_pypi_req(
+    req: PipRequirement,
+    requirements_file: PipRequirementsFile,
+    require_hashes: bool,
+    pip_deps_dir: RootedPath,
+    allow_binary: bool,
+) -> list[dict[str, Any]]:
+    download_infos: list = []
+
+    def _process_req_wrapper(
+        DPIs: list[DistributionPackageInfo], log_string: str
+    ) -> Optional[dict[str, Any]]:
+        for dpi in DPIs:
+            log.info(log_string, dpi.path.name)
+            info: dict[str, Any] = _process_req(
+                req, requirements_file, require_hashes, pip_deps_dir, dpi.download_info, dpi=dpi
+            )
+            return info
+        return None
+
+    artifacts: list[DistributionPackageInfo] = _process_package_distributions(
+        req, pip_deps_dir, allow_binary
+    )
+    artifacts_to_download: list[DistributionPackageInfo] = [a for a in artifacts if a.path.exists()]
+    files: dict[str, Union[str, PathLike[str]]] = {
+        dpi.url: dpi.path for dpi in artifacts_to_download
+    }
+    asyncio.run(async_download_files(files, get_config().concurrency_limit))
```

Comment:
Shouldn't we download all wheels from requirement.txt at once?

Reply:
Previously that was the case, yes, so in that regard this is change introduces a regression in the behaviour. Downloading everything in a single async call can be a double-edged sword though. Unless we're seeing a noticeable performance penalty I'd argue that downloading from pypi on a req by req basis gives us the possibility to either recover from errors or fail fatally early without having to potentially download hundreds of MB of data first.